[CI] FloatHierarchicalKMeansTests testSoarAssignmentsValidAfterEmptyClusterRemoval failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #12600 / ubuntu-2404_platform-support-unix](https://gradle-enterprise.elastic.co/s/agbh52yot4xmg)
- [elasticsearch-periodic-platform-support #12600 / ubuntu-2404_platform-support-unix](https://gradle-enterprise.elastic.co/s/ng3fhlntvuwcy)

**Reproduction Line:**
```
./gradlew ":server:test" --tests "org.elasticsearch.index.codec.vectors.cluster.FloatHierarchicalKMeansTests.testSoarAssignmentsValidAfterEmptyClusterRemoval" -Dtests.seed=58DDB35C8C17CF3 -Dtests.locale=nnh -Dtests.timezone=Europe/Kyiv -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.index.codec.vectors.cluster.FloatHierarchicalKMeansTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testSoarAssignmentsValidAfterEmptyClusterRemoval'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: Failed to assign soar vector to centroid
```

**Issue Reasons:**
- [main] 2 consecutive failures in step ubuntu-2404_platform-support-unix
- [main] 1 consecutive failures in pipeline elasticsearch-periodic-platform-support
- [main] 2 failures in test testSoarAssignmentsValidAfterEmptyClusterRemoval (0.7% fail rate in 298 executions)
- [main] 2 failures in step ubuntu-2404_platform-support-unix (100.0% fail rate in 2 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-core-infra.

:eyes:

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -595,9 +595,6 @@ tests:
 - class: org.elasticsearch.xpack.esql.qa.single_node.EsqlSpecApproximationBucketIncludeEmptyBucketsIT
   method: test {csv-spec:bucket_include_empty_buckets.Large number of buckets doesn't OOM - stats}
   issue: https://github.com/elastic/elasticsearch/issues/155921
-- class: org.elasticsearch.index.codec.vectors.cluster.FloatHierarchicalKMeansTests
-  method: testSoarAssignmentsValidAfterEmptyClusterRemoval
-  issue: https://github.com/elastic/elasticsearch/issues/155931
 - class: org.elasticsearch.test.rest.ClientYamlTestSuiteIT
   method: test {yaml=search.highlight/70_number_of_fragments/number_of_fragments at the maximum should SUCCEED}
   issue: https://github.com/elastic/elasticsearch/issues/155969

--- a/server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/Soar.java
+++ b/server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/Soar.java
@@ -202,6 +202,10 @@ public abstract class Soar<V> {
         float[] distances,
         CentroidOps<V> ops
     ) {
+        if (centroidCount <= 0) {
+            // no other centroid available to spill to (e.g. an empty neighborhood after empty-cluster removal)
+            return NO_SOAR_ASSIGNMENT;
+        }
         V currentCentroid = centroids[currAssignment];
         float vectorCentroidDist = ops.squareDistance(vector, currentCentroid);
         if (vectorCentroidDist <= SOAR_MIN_DISTANCE) {

